File: core/src/main/cpp/crypto/MslSignatureEnvelope.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <crypto/MslSignatureEnvelope.h>
#include <io/MslEncoderException.h>
#include <io/MslEncoderFactory.h>
#include <io/MslEncoderFormat.h>
#include <io/MslObject.h>
#include <io/MslVariant.h>
#include <MslCryptoException.h>
#include <MslEncodingException.h>
#include <MslError.h>
#include <MslInternalException.h>
#include <util/Base64.h>

using namespace std;
using netflix::msl::io::MslEncoderException;
using netflix::msl::io::MslEncoderFactory;
using netflix::msl::io::MslEncoderFormat;
using netflix::msl::io::MslObject;
using netflix::msl::io::Variant;

namespace netflix {
namespace msl {
namespace crypto {

namespace // anonymous
{

/** Key version. */
const std::string KEY_VERSION = "version";
/** Key algorithm. */
const std::string KEY_ALGORITHM = "algorithm";
/** Key signature. */
const std::string KEY_SIGNATURE = "signature";

} //namespace anonymous

// --- start MslSignatureEnvelope::Version

const MslSignatureEnvelope::Version MslSignatureEnvelope::Version::V1(MslSignatureEnvelope::Version::v1, "V1");
const MslSignatureEnvelope::Version MslSignatureEnvelope::Version::V2(MslSignatureEnvelope::Version::v2, "V2");
const MslSignatureEnvelope::Version MslSignatureEnvelope::Version::INVALID(MslSignatureEnvelope::Version::invalid, "INVALID");

// static
const std::vector<MslSignatureEnvelope::Version>& MslSignatureEnvelope::Version::getValues()
{
    static std::vector<Version> gValues;
    if (gValues.empty()) {
        gValues.push_back(V1);
        gValues.push_back(V2);
    }
    return gValues;
}

// static
MslSignatureEnvelope::Version MslSignatureEnvelope::Version::valueOf(int version)
{
    switch (version) {
        case 1: return V1;
        case 2: return V2;
        default: throw IllegalArgumentException("Unknown signature envelope version.");
    }
}

// --- end MslSignatureEnvelope::Version

shared_ptr<ByteArray> MslSignatureEnvelope::getBytes(shared_ptr<MslEncoderFactory> encoder, const MslEncoderFormat& format) const
{
    if (version_ == Version::V1) {
        return signature_;
    }
    else if (version_ == Version::V2) {
        std::shared_ptr<MslObject> mo = encoder->createObject(std::map<std::string, Variant>());
        mo->put<int>(KEY_VERSION, version_);
        mo->put<std::string>(KEY_ALGORITHM, algorithm_.stringValue());
        mo->put<std::shared_ptr<ByteArray>>(KEY_SIGNATURE, signature_);
        return encoder->encodeObject(mo, format);
    }
    else {
        std::stringstream ss;
        ss << "Signature envelope version " << version_ << " encoding unsupported.";
        throw MslInternalException(ss.str());
    }
}

// static
MslSignatureEnvelope MslSignatureEnvelope::parse(shared_ptr<ByteArray> envelope,
        shared_ptr<MslEncoderFactory> encoder, const Version& version)
{
    // Parse envelope.
    if (version == Version::V1) {
        return MslSignatureEnvelope(envelope);
    }
    else if (version == Version::V2) {
        try {
            // We expect the byte representation to be a MSL object.
            shared_ptr<MslObject> envelopeMo = encoder->parseObject(envelope);

            // Verify version.
            try {
                const Version v = Version::valueOf(envelopeMo->getInt(KEY_VERSION));
                if (v != Version::V2) {
                    std::stringstream ss;
                    ss << "signature envelope " << envelopeMo;
                    throw MslCryptoException(MslError::UNSUPPORTED_SIGNATURE_ENVELOPE, ss.str());
                }
            } catch (const IllegalArgumentException& e) {
                std::stringstream ss;
                ss << "signature envelope " << envelopeMo;
                throw MslCryptoException(MslError::UNIDENTIFIED_SIGNATURE_ENVELOPE, ss.str(), e);
            }

            // Grab algorithm.
            MslConstants::SignatureAlgo algorithm;
            try {
                algorithm = MslConstants::SignatureAlgo::fromString(envelopeMo->getString(KEY_ALGORITHM));
            } catch (const IllegalArgumentException& e) {
                std::stringstream ss;
                ss << "signature envelope " << envelopeMo;
                throw MslCryptoException(MslError::UNIDENTIFIED_ALGORITHM, ss.str(), e);
            }

            // Grab signature.
            shared_ptr<ByteArray> signature = envelopeMo->getBytes(KEY_SIGNATURE);

            // Return the envelope.
            return MslSignatureEnvelope(algorithm, signature);
        } catch (const MslEncoderException& e) {
            std::stringstream ss;
            ss << "signature envelope " << util::Base64::encode(envelope);
            throw MslEncodingException(MslError::MSL_PARSE_ERROR, ss.str(), e);
        }
    }
    else {
        std::stringstream ss;
        ss << "signature envelope " << util::Base64::encode(envelope);
        throw MslCryptoException(MslError::UNSUPPORTED_SIGNATURE_ENVELOPE, ss.str());
    }
}

//static
MslSignatureEnvelope MslSignatureEnvelope::parse(shared_ptr<ByteArray> envelope, shared_ptr<MslEncoderFactory> encoder)
{
    // Attempt to convert this to a MSL object.
    shared_ptr<MslObject> envelopeMo = encoder->createObject();
    try {
        // If this is a MSL object, we expect the byte representation to be
        // decodable.
        envelopeMo = encoder->parseObject(envelope);
    } catch (const MslEncoderException& e) {
        // do nothing, leave envelopeMo empty.
    }

    // Determine the envelope version.
    //
    // If there is no MSL object, or there is no version field (as the
    // binary signature may coincidentally parse into a MSL object), then
    // this is a version 1 envelope.
    Version version;
    if (!envelopeMo || !envelopeMo->has(KEY_VERSION)) {
        version = Version::V1;
    } else {
        try {
            version = Version::valueOf(envelopeMo->getInt(KEY_VERSION));
        } catch (const MslEncoderException& e) {
            // There is a possibility that this is a version 1 envelope.
            version = Version::V1;
        } catch (const IllegalArgumentException& e) {
            // There is a possibility that this is a version 1 envelope.
            version = Version::V1;
        }
    }

    // FIXME: Why not just call parse(envelope, encoder, version) here?

    // Parse envelope.
    if (version == Version::V1) {
        return MslSignatureEnvelope(envelope);
    }
    else if (version == Version::V2) {
        try {
            const std::string algoStr = envelopeMo->getString(KEY_ALGORITHM);
            const MslConstants::SignatureAlgo algorithm = MslConstants::SignatureAlgo::fromString(algoStr);
            shared_ptr<ByteArray> signature = envelopeMo->getBytes(KEY_SIGNATURE);
            return MslSignatureEnvelope(algorithm, signature);
        } catch (const MslEncoderException& e) {
            // It is extremely unlikely but possible that this is a
            // version 1 envelope.
            return MslSignatureEnvelope(envelope);
        } catch (const IllegalArgumentException& e) {
            // It is extremely unlikely but possible that this is a
            // version 1 envelope.
            return MslSignatureEnvelope(envelope);
        }
    }
    else {
        std::stringstream ss;
        ss << "signature envelope " << util::Base64::encode(envelope);
        throw MslCryptoException(MslError::UNSUPPORTED_SIGNATURE_ENVELOPE, ss.str());
    }
}

}}} // namespace netflix::msl::crypto

File: core/src/main/javascript/crypto/MslSignatureEnvelope.js
/**
 * Copyright (c) 2013-2018 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * <p>MSL signature envelopes contain all of the information necessary for
 * verifying data using a known key.</p>
 *
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
	"use strict";
	
	var Class = require("../util/Class.js");
	var AsyncExecutor = require('../util/AsyncExecutor.js');
	var MslInternalException = require('../MslInternalException.js');
	var MslCryptoException = require('../MslCryptoException.js');
	var MslConstants = require('../MslConstants.js');
	var MslEncoderException = require('../io/MslEncoderException.js');
	var MslEncodingException = require('../MslEncodingException.js');
	var MslError = require('../MslError.js');
	var Base64 = require('../util/Base64.js');
	
    /**
     * Key version.
     * @const
     * @type {string}
     */
    var KEY_VERSION = "version";
    /**
     * Key algorithm.
     * @const
     * @type {string}
     */
    var KEY_ALGORITHM = "algorithm";
    /**
     * Key signature.
     * @const
     * @type {string}
     */
    var KEY_SIGNATURE = "signature";

    /** Versions. */
    var Version = {
        /**
         * <p>Version 1.</p>
         *
         * {@code signature}
         *
         * <p>The signature is represented as raw bytes.</p>
         */
        V1 : 1,
        /**
         * <p>Version 2.</p>
         *
         * {@code {
         *   "#mandatory" : [ "version", "algorithm", "signature" ],
         *   "version" : "number",
         *   "algorithm" : "string",
         *   "signature" : "binary"
         * }} where:
         * <ul>
         * <li>{@code version} is the number '2'</li>
         * <li>{@code algorithm} is one of the recognized signature algorithms</li>
         * <li>{@code signature} is the signature</li>
         * </ul>
         *
         * <p>Supported algorithms:
         * <table>
         * <tr><th>Algorithm</th><th>Description</th>
         * <tr><td>HmacSHA256</td><td>HMAC w/SHA-256</td></tr>
         * <tr><td>SHA256withRSA</td><td>RSA signature w/SHA-256</td></tr>
         * <tr><td>AESCmac</td><td>AES CMAC</td></tr>
         * </table></p>
         */
        V2 : 2,
    };

    var MslSignatureEnvelope = module.exports = Class.create({
        /**
         * Create a new signature envelope with the provided data.
         *
         * @param {Version} version the envelope version.
         * @param {?MslConstants$SignatureAlgo} algorithm the signature algorithm. May be null.
         * @param {Uint8Array} signature the signature.
         */
        init: function init(version, algorithm, signature) {
            // The properties.
            var props = {
                version: { value: version, writable: false, enumerable: false, configurable: false },
                algorithm: { value: algorithm, writable: false, configurable: false },
                signature: { value: signature, writable: false, configurable: false },
            };
            Object.defineProperties(this, props);
        },
        
        /**
         * Returns the signature envelope in byte form.
         * 
         * @param {MslEncoderFactory} encoder MSL encoder factory.
         * @param {MslEncoderFormat} format MSL encoder format.
         * @param {{result: function(Uint8Array), error: function(Error)}}
         *        callback the callback functions that will receive the byte
         *        representation of the signature envelope or any thrown
         *        exceptions.
         * @throws MslEncoderException if there is an error encoding the envelope.
         * @throws MslInternalException if the envelope version is not supported.
         */
        getBytes: function getBytes(encoder, format, callback) {
            AsyncExecutor(callback, function() {
                switch (this.version) {
                    case Version.V1:
                        return this.signature;
                    case Version.V2:
                        var mo = encoder.createObject();
                        mo.put(KEY_VERSION, this.version);
                        mo.put(KEY_ALGORITHM, this.algorithm);
                        mo.put(KEY_SIGNATURE, this.signature);
                        encoder.encodeObject(mo, format, callback);
                        break;
                    default:
                        throw new MslInternalException("Signature envelope version " + this.version + " encoding unsupported.");
                }
            }, this);
        },
    });

    /**
     * <p>This method has two acceptable parameter lists.</p>
     *
     * <p>The first form creates a version 1 signature envelope.</p>
     *
     * @param {Uint8Array} signature the signature.
     * @param {{result: function(MslSignatureEnvelope), error: function(Error)}}
     *        callback the callback functions that will receive the envelope
     *        or any thrown exceptions.
     *
     * <p>The second form creates a version 2 signature envelope.</p>
     *
     * @param {MslConstants$SignatureAlgo} algorithm the signature algorithm.
     * @param {Uint8Array} signature the signature.
     * @param {{result: function(MslSignatureEnvelope), error: function(Error)}}
     *        callback the callback functions that will receive the envelope
     *        or any thrown exceptions.
     */
    var MslSignatureEnvelope$create = function MslSignatureEnvelope$create(/* variable arguments */) {
        var version,
            signature,
            algorithm,
            callback;

        // Handle the first form.
        if (arguments.length == 2) {
            version = Version.V1;
            signature = arguments[0];
            algorithm = null;
            callback = arguments[1];
        }

        // Handle the second form.
        else if (arguments.length == 3) {
            version = Version.V2;
            algorithm = arguments[0];
            signature = arguments[1];
            callback = arguments[2];
        }

        // Malformed arguments are not explicitly handled, just as with any
        // other function.

        AsyncExecutor(callback, function() {
            return new MslSignatureEnvelope(version, algorithm, signature);
        });
    };

    /**
     * Create a new signature envelope from the provided envelope bytes. If a
     * signature version is provided then the MSL object is parsed accordingly.
     *
     * @param {Uint8Array} envelope the raw envelope bytes.
     * @param {?Version} version the envelope version.
     *        May be null.
     * @param {MslEncoderFactory} MSL encoder factory.
     * @param {{result: function(MslSignatureEnvelope), error: function(Error)}}
     *        callback the callback functions that will receive the envelope
     *        or any thrown exceptions.
     * @throws MslCryptoException if there is an error processing the signature
     *         envelope.
     * @throws MslEncodingException if there is an error parsing the envelope.
     * @see #getBytes(MslEncoderFactory, MslEncoderFormat)
     */
    var MslSignatureEnvelope$parse = function MslSignatureEnvelope$parse(envelope, version, encoder, callback) {
        AsyncExecutor(callback, function() {
            var algorithm, signature;
            var envelopeMo;
            
            if (version) {
                switch (version) {
                    case Version.V1:
                        return new MslSignatureEnvelope(Version.V1, null, envelope);
                    case Version.V2:
                        try {
                            // We expect the byte representation to be a MSL object.
                            envelopeMo = encoder.parseObject(envelope);

                            // Verify version.
                            var v = envelopeMo.getInt(KEY_VERSION);
                            if (Version.V2 != v)
                                throw new MslCryptoException(MslError.UNSUPPORTED_SIGNATURE_ENVELOPE, "signature envelope " + envelope);

                            // Grab algorithm.
                            algorithm = MslConstants.SignatureAlgo.fromString(envelopeMo.getString(KEY_ALGORITHM));
                            if (!algorithm)
                                throw new MslCryptoException(MslError.UNIDENTIFIED_ALGORITHM, "signature envelope " + envelope);

                            // Grab signature.
                            signature = envelopeMo.getBytes(KEY_SIGNATURE);

                            // Return the envelope.
                            return new MslSignatureEnvelope(Version.V2, algorithm, signature);
                        } catch (e) {
                            if (e instanceof MslEncoderException)
                                throw new MslEncodingException(MslError.MSL_PARSE_ERROR, "signature envelope " + envelope, e);
                            throw e;
                        }
                    default:
                        throw new MslCryptoException(MslError.UNSUPPORTED_SIGNATURE_ENVELOPE, "signature envelope " + Base64.encode(envelope));
                }
            }

            // Attempt to convert this to a MSL object.
            try {
                // If this is a MSL object, we expect the byte representation to be
                // decodable.
                envelopeMo = encoder.parseObject(envelope);
            } catch (e) {
                if (e instanceof MslEncoderException)
                    envelopeMo = null;
                else
                    throw e;
            }

            // Determine the envelope version.
            //
            // If there is no MSL object, or there is no version field (as the
            // binary signature may coincidentally parse into a MSL object), then
            // this is a version 1 envelope.
            var envelopeVersion;
            if (!envelopeMo || !envelopeMo.has(KEY_VERSION)) {
                envelopeVersion = Version.V1;
            } else {
                try {
                    envelopeVersion = envelopeMo.getInt(KEY_VERSION);
                    if (envelopeVersion !== envelopeVersion) {
                        // There is a possibility that this is a version 1 envelope.
                        envelopeVersion = Version.V1;
                    }
                    var recognized = false;
                    for (var ver in Version) {
                        if (Version[ver] == envelopeVersion) {
                            recognized = true;
                            break;
                        }
                    }
                    if (!recognized) {
                        // There is a possibility that this is a version 1 envelope.
                        envelopeVersion = Version.V1;
                    }
                } catch (e) {
                    if (e instanceof MslEncoderException) {
                        // There is a possibility that this is a version 1 envelope.
                        envelopeVersion = Version.V1;
                    }
                }
            }

            // Parse envelope.
            switch (envelopeVersion) {
                case Version.V1:
                    return new MslSignatureEnvelope(envelopeVersion, null, envelope);
                case Version.V2:
                    try {
                        algorithm = MslConstants.SignatureAlgo.fromString(envelopeMo.getString(KEY_ALGORITHM));
                        signature = envelopeMo.getBytes(KEY_SIGNATURE);

                        // Verify algorithm.
                        if (!algorithm) {
                            // It is extremely unlikely but possible that this is a
                            // version 1 envelope.
                            return new MslSignatureEnvelope(Version.V1, null, envelope);
                        }
                        return new MslSignatureEnvelope(envelopeVersion, algorithm, signature);
                    } catch (e) {
                        if (e instanceof MslEncoderException) {
                            // It is extremely unlikely but possible that this is a
                            // version 1 envelope.
                            return new MslSignatureEnvelope(Version.V1, null, envelope);
                        }
                        throw e;
                    }
                default:
                    throw new MslCryptoException(MslError.UNSUPPORTED_SIGNATURE_ENVELOPE, "signature envelope " + envelope);
            }
        });
    };
    
    // Exports.
    module.exports.create = MslSignatureEnvelope$create;
    module.exports.parse = MslSignatureEnvelope$parse;
    module.exports.Version = Version;
})(require, (typeof module !== 'undefined') ? module : mkmodule('MslSignatureEnvelope'));